README.md
# Rust-ConfigParser [![CI](https://github.com/Basicprogrammer10/Rust-ConfigParser/actions/workflows/main.yml/badge.svg)](https://github.com/Basicprogrammer10/Rust-ConfigParser/actions/workflows/main.yml) ![Crates.io](https://img.shields.io/crates/d/simple_config_parser) ![Lines of code](https://img.shields.io/tokei/lines/github/Basicprogrammer10/Rust-ConfigParser)
⚙ Very simple config parsing lib for rust!

I made this because I just needed a simple config parser for one of my projects and wanted to learn how to make a rust crate. Hopefully you will find it useful as well. :P

## 💠 Install

Just add the following to your `Cargo.toml`:
```toml
[dependencies]
simple_config_parser = "1.0.0"
```

## 📀 Quick Start

This config parser is made for use with a simplified version of an ini file. There are no sections and currently no Escape character support.
```ini
; This is a comment
# This is also a comment
hello = World
rust = Is great
test = "TEST"
```

## 🐳 Why

There are already a few config parsers out there for rust so why use this one?

There are a few reasons:
- It's super simple to use
- Its faster then some other popular config parsers (by only a few hundred Nano seconds but still)
- It's code is easy to understand (For me at least)
- It would make me happy (:P)

## 💥 Examples

Create a new config from text and a file.
```rust
// Import Lib
use simple_config_parser::Config;

// Create a new config and parse text
let cfg = Config::new()
    .text("hello = world")
    .unwrap();

// Create a new config from a file
let cfg2 = Config::new()
    .file("config.cfg")
    .unwrap();
```

Get a value from a config.
```rust
// Import Lib
use simple_config_parser::Config;

// Create a new config with no file
let cfg = Config::new()
    .text("hello = World\nrust = Is great")
    .unwrap();

// Get a value from the config (As a string)
println!("Hello, {}", cfg.get_str("hello").unwrap());
```

Get value from a config as any type that implements FromStr.
```rust
// Import Lib
use simple_config_parser::Config;

// Create a new config with no file
let mut cfg = Config::new()
    .text("hello = true\nrust = 15\npi = 3.1415926535")
    .unwrap();

// Get a value from the config as bool
assert_eq!(cfg.get::<bool>("hello").unwrap(), true);

// Get a value from the config as int
assert_eq!(cfg.get::<i32>("rust").unwrap(), 15);

// Get a value from the config as float
assert_eq!(cfg.get::<f32>("pi").unwrap(), 3.1415926535);
```


changelog.md
# Changelog
## 1.0.0
- Cleanup Syntax of Loading / Reading Config
- Remove all get_(bool, int, float)
- Add get_str for getting raw config value
- add get<T> for gatting a config value as any type (that impls FromStr)
- Update Exmaples
- Cleanup Code

## 0.1.6
- Add docs for internal functions
- Update some formatting in the readme
- Make the parsed config data available to the user

## 0.1.5
- Update Keywords to be valid for [Crates.io](crates.io)
- Update Docs for some functions
- Added Readme to the `lib.rs` file
- Add `get_bool` function to get config values as Booleans
- Add `get_int` function to get config values as Integers
- Add `get_float` function to get config values as Floats

## 0.1.4
 - Add Keywords, Categories and documentation to Cargo.toml
 - Allow for comments mid line
 - Remove Trailing / Leading spaces in Key and Value
 - Use more constants in the code
 - Ignore Sections for the time being
 - Add Benchmarks
 - Most recently defined config options will be used first
    - Ex: If you have a config file that defines the key hello twice the one lower in the file will be used
- Make `cfg.get(key)` return a `Option<String>` instead of a `String`


tests/test.rs
use simple_config_parser::Config;

#[test]
/// Test parsing a config string
fn test_config_from_string() {
    let cfg = Config::new()
        .text("hello = world\nrust = is great\ntest = \"TEST\"")
        .unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "world");
    assert_eq!(cfg.get_str("rust").unwrap(), "is great");
    assert_eq!(cfg.get_str("test").unwrap(), "\"TEST\"");
}

#[test]
/// Test loading and parsing a config file
fn test_config_from_file() {
    let cfg = Config::new().file("config.cfg").unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "World");
    assert_eq!(cfg.get_str("rust").unwrap(), "Is great");
    assert_eq!(cfg.get_str("test").unwrap(), "\"TEST\"");
}

#[test]
#[should_panic]
/// Test loading and parsing a config file with comments
fn test_config_from_string_with_comments() {
    let cfg = Config::new()
        .text("hello = world\n#rust = is great\n;test = \"TEST\"")
        .unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "world");

    // This should panic
    cfg.get_str("rust").unwrap();
    cfg.get_str("test").unwrap();
}

#[test]
/// Test case sensitivity of config keys
fn test_case_sensitivity() {
    let cfg = Config::new()
        .text("HeLlO = world\nRuSt = is great\nTeSt = \"TEST\"")
        .unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "world");
    assert_eq!(cfg.get_str("RUST").unwrap(), "is great");
    assert_eq!(cfg.get_str("tEsT").unwrap(), "\"TEST\"");
}

#[test]
/// Test comments in a config file not at the beginning of a line
fn test_comments_in_file() {
    let cfg = Config::new()
        .text("hello = world\nrust = is great # Comment\ntest = \"TEST\"; Nother Comment")
        .unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "world");
    assert_eq!(cfg.get_str("RUST").unwrap(), "is great");
    assert_eq!(cfg.get_str("tEsT").unwrap(), "\"TEST\"");
}

#[test]
/// Test parseing will Ignore Sections
fn test_ignoring_sections() {
    let cfg = Config::new()
        .text("[section]\nhello = world\n[section2]\nrust = is great\n[section3]\ntest = \"TEST\"")
        .unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "world");
    assert_eq!(cfg.get_str("rust").unwrap(), "is great");
    assert_eq!(cfg.get_str("test").unwrap(), "\"TEST\"");
}

#[test]
/// The most recently defined key will be used
fn test_last_key_wins() {
    let cfg = Config::new().text("hello = people\nhello = world").unwrap();

    assert_eq!(cfg.get_str("hello").unwrap(), "world");
}

#[test]
/// Test getting value as a bool
fn test_bool_value() {
    let cfg = Config::new().text("test = true\ntset = false").unwrap();

    assert_eq!(cfg.get::<bool>("test").unwrap(), true);
    assert_eq!(cfg.get::<bool>("tset").unwrap(), false);
}

#[test]
/// Test getting value as a integer
fn test_int_value() {
    let cfg = Config::new().text("test = 1\ntset = 100").unwrap();

    assert_eq!(cfg.get::<i32>("test").unwrap(), 1);
    assert_eq!(cfg.get::<i32>("tset").unwrap(), 100);
}

#[test]
/// Test getting value as a float
fn test_float_value() {
    let cfg = Config::new()
        .text("test = 1.123\npi = 3.14159 ; And so on")
        .unwrap();

    assert_eq!(cfg.get::<f32>("test").unwrap(), 1.123);
    assert_eq!(cfg.get::<f32>("pi").unwrap(), 3.14159);
}

#[test]
/// Test the priority of loading data more than one time
fn test_load_priority() {
    let cfg = Config::new()
        .text("a = 1\nb = 2")
        .unwrap()
        .text("a = 2\nb = 4")
        .unwrap();

    assert_eq!(cfg.get_str("a").unwrap(), "2");
    assert_eq!(cfg.get_str("b").unwrap(), "4");
}


testing/src/main.rs
use simple_config_parser;

fn main() {
    let mut cfg = simple_config_parser::Config::new(None);
    println!("Hello, world!");
}


src/config.rs
//! This module contains the things needed to load and parse ini like configuration files
use std::fs;
use std::path::Path;

/// Define valid comment chars.
const COMMENT_CHARS: [&str; 2] = ["#", ";"];

/// Config Struct
pub struct Config {
    /// Raw Data of the Config
    pub data: Vec<[String; 2]>,
}

/// Some errors that can be thrown by this module
#[derive(Debug)]
pub enum ConfigError {
    /// Error reading the file from disk
    /// Could have been caused by the file not existing or being inaccessible.
    FileReadError,
    /// File path has not been defined
    /// You need to define the path to the config file before using this function.
    /// Or just use `cfg.parse("<STRING>");` instead.
    NoFileDefined,
    /// The config data is not valid
    /// The data read from the file is not valid.
    InvalidConfig,
    /// Error Parseing config value into specified type
    ParseError,
    /// No item for the key provided exists
    NoItem,
}

/// Removes any comments from each line of the config file.
fn remove_comments(str: String) -> String {
    let mut s = str.as_str();
    for i in COMMENT_CHARS.iter() {
        s = s.split(i).next().unwrap();
    }
    s.to_string()
}

/// Config Implementation
impl Config {
    /// Create a new Config struct
    /// ## Example
    /// ```rust
    /// // Import Lib
    /// use simple_config_parser::Config;
    ///
    /// let mut cfg = Config::new();
    /// ```
    pub fn new() -> Self {
        Config { data: Vec::new() }
    }

    /// Reads and parses config from a file
    ///
    /// If called more than one time it will append the current values.
    /// So the recently appended valued will take priority
    /// ## Example
    /// ```rust
    /// // Import Lib
    /// use simple_config_parser::Config;
    ///
    /// // Create a new config with a file
    /// let mut cfg = Config::new().file("config.cfg").unwrap();
    ///
    /// // Read a value
    /// assert_eq!(cfg.get_str("hello").unwrap(), "World");
    /// ```
    pub fn file<T>(self, file: T) -> Result<Self, ConfigError>
    where
        T: AsRef<Path>,
    {
        let contents = match fs::read_to_string(file) {
            Ok(contents) => contents,
            Err(_) => return Err(ConfigError::FileReadError),
        };

        let mut data = self.data;
        data.append(&mut Config::parse(contents)?);

        Ok(Self { data })
    }

    /// Parses config from text or anything that impls fmt::Display
    /// ## Example
    /// ```rust
    /// // Import Lib
    /// use simple_config_parser::Config;
    ///
    /// // Create a new config with text
    /// let mut cfg = Config::new().text("hello = World").unwrap();
    ///
    /// // Read a value
    /// assert_eq!(cfg.get_str("hello").unwrap(), "World");
    /// ```
    pub fn text<T>(self, text: T) -> Result<Self, ConfigError>
    where
        T: std::fmt::Display,
    {
        let data = Config::parse(text.to_string())?;

        Ok(Self { data })
    }

    /// Get a value from config as ayn type (That Impls str::FromStr)
    /// ## Example
    /// ```rust
    /// // Import Lib
    /// use simple_config_parser::Config;
    ///
    /// // Create a new config with text
    /// // I knew a lot more of pi at one point :P
    /// let mut cfg = Config::new().text("pi = 3.14159265358979").unwrap();
    ///
    /// // Read a value
    /// assert_eq!(cfg.get::<f32>("pi").unwrap(), 3.14159265358979);
    /// ```
    pub fn get<T>(&self, key: &str) -> Result<T, ConfigError>
    where
        T: core::str::FromStr,
    {
        let key = key.to_string().to_lowercase();
        for i in self.data.iter().rev() {
            if i[0] != key {
                continue;
            }
            match i[1].parse() {
                Ok(i) => return Ok(i),
                Err(_) => return Err(ConfigError::ParseError),
            }
        }

        Err(ConfigError::NoItem)
    }

    /// Get a value from config as a String
    /// ## Example
    /// ```rust
    /// // Import Lib
    /// use simple_config_parser::Config;
    ///
    /// // Create a new config with text
    /// let mut cfg = Config::new().text("pi = 3.14159265358979").unwrap();
    ///
    /// // Read a value
    /// assert_eq!(cfg.get_str("pi").unwrap(), "3.14159265358979");
    /// ```
    pub fn get_str(&self, key: &str) -> Result<String, ConfigError> {
        let key = key.to_string().to_lowercase();
        for i in self.data.iter().rev() {
            if i[0] != key {
                continue;
            }
            return Ok(i[1].to_string());
        }

        Err(ConfigError::NoItem)
    }

    /// Parse a string into the config
    fn parse(input_data: String) -> Result<Vec<[String; 2]>, ConfigError> {
        let mut done: Vec<[String; 2]> = Vec::new();

        for line in input_data.lines() {
            // Remove any space at the beginning of the line
            let mut line = line.trim().to_string();

            // Skip empty / commented lines and sections (for now)
            match line.chars().next() {
                Some(i) if COMMENT_CHARS.contains(&&i.to_string()[..]) => continue,
                Some('[') => continue,
                Some(_) => {}
                None => continue,
            }

            // Remove any comments from the line
            line = remove_comments(line.to_string());

            // Split the line into key and value
            let parts: Vec<&str> = line.split('=').collect();
            if parts.len() != 2 {
                return Err(ConfigError::InvalidConfig);
            }

            // Remove any spaces in the key
            let key = parts[0].replace(" ", "").to_lowercase();
            let value = parts[1].trim().to_string();

            done.push([key, value]);
        }

        Ok(done)
    }
}

impl Default for Config {
    fn default() -> Config {
        Config::new()
    }
}


benches/bench.rs
#![feature(test)]

extern crate test;
use test::Bencher;

use simple_config_parser::Config;

pub fn parse_string() -> Option<()> {
    let cfg = Config::new().text("hello = world ; Comment").ok()?;

    Some(())
}

pub fn parse_string_get() -> Option<()> {
    let cfg = Config::new().text("hello = world ; Comment").ok()?;
    cfg.get_str("hello").unwrap();
    Some(())
}

pub fn parse_messy_string() -> Option<()> {
    let cfg = Config::new()
        .text("     hello   =   world ;#;#;#; Comment")
        .ok()?;

    Some(())
}

pub fn parse_file() -> Option<()> {
    let cfg = Config::new().file("config.cfg").ok()?;

    Some(())
}

pub fn parse_string_get_bool() -> Option<()> {
    let cfg = Config::new().text("hello = true ; Comment").ok()?;
    cfg.get::<bool>("hello").ok()?;

    Some(())
}

pub fn parse_string_get_int() -> Option<()> {
    let cfg = Config::new().text("hello = 1 ; Comment").ok()?;
    cfg.get::<i32>("hello").ok()?;

    Some(())
}

pub fn parse_string_get_float() -> Option<()> {
    let mut cfg = Config::new().text("hello = 1.0 ; Comment").ok()?;
    cfg.get::<f32>("hello").ok()?;

    Some(())
}

#[bench]
/// Basic config parsing benchmark.
fn bench_parse(b: &mut Bencher) {
    b.iter(|| parse_string());
}

#[bench]
/// Parse string and get value.
fn bench_parse_get(b: &mut Bencher) {
    b.iter(|| parse_string_get());
}

#[bench]
/// Parse messy config string
fn bench_parse_messy(b: &mut Bencher) {
    b.iter(|| parse_messy_string());
}

#[bench]
/// Config file reading and parsing benchmark.
fn bench_read_parse(b: &mut Bencher) {
    b.iter(|| parse_file());
}

#[bench]
/// Parse string and get bool value.
fn bench_parse_get_bool(b: &mut Bencher) {
    b.iter(|| parse_string_get_bool());
}

#[bench]
/// Parse string and get bool value.
fn bench_parse_get_int(b: &mut Bencher) {
    b.iter(|| parse_string_get_int());
}

#[bench]
/// Parse string and get bool value.
fn bench_parse_get_float(b: &mut Bencher) {
    b.iter(|| parse_string_get_float());
}


src/lib.rs
/*!
# Rust-ConfigParser [![CI](https://github.com/Basicprogrammer10/Rust-ConfigParser/actions/workflows/main.yml/badge.svg)](https://github.com/Basicprogrammer10/Rust-ConfigParser/actions/workflows/main.yml) ![Crates.io](https://img.shields.io/crates/d/simple_config_parser) ![Lines of code](https://img.shields.io/tokei/lines/github/Basicprogrammer10/Rust-ConfigParser)
⚙ Very simple config parsing lib for rust!

I made this because I just needed a simple config parser for one of my projects and wanted to learn how to make a rust crate. Hopefully you will find it useful as well. :P

## 💠 Install

Just add the following to your `Cargo.toml`:
```toml
[dependencies]
simple_config_parser = "1.0.0"
```

## 📀 Quick Start

This config parser is made for use with a simplified version of an ini file. There are no sections and currently no Escape character support.
```ini
; This is a comment
# This is also a comment
hello = World
rust = Is great
test = "TEST"
```

## 🐳 Why

There are already a few config parsers out there for rust so why use this one?

There are a few reasons:
- It's super simple to use
- Its faster then some other popular config parsers (by only a few hundred Nano seconds but still)
- It's code is easy to understand (For me at least)
- It would make me happy (:P)

## 💥 Examples

Create a new config from text and a file.
```rust
// Import Lib
use simple_config_parser::Config;

// Create a new config and parse text
let cfg = Config::new()
    .text("hello = world")
    .unwrap();

// Create a new config from a file
let cfg2 = Config::new()
    .file("config.cfg")
    .unwrap();
```

Get a value from a config.
```rust
// Import Lib
use simple_config_parser::Config;

// Create a new config with no file
let cfg = Config::new()
    .text("hello = World\nrust = Is great")
    .unwrap();

// Get a value from the config (As a string)
println!("Hello, {}", cfg.get_str("hello").unwrap());
```

Get value from a config as any type that implements FromStr.
```rust
// Import Lib
use simple_config_parser::Config;

// Create a new config with no file
let mut cfg = Config::new()
    .text("hello = true\nrust = 15\npi = 3.1415926535")
    .unwrap();

// Get a value from the config as bool
assert_eq!(cfg.get::<bool>("hello").unwrap(), true);

// Get a value from the config as int
assert_eq!(cfg.get::<i32>("rust").unwrap(), 15);

// Get a value from the config as float
assert_eq!(cfg.get::<f32>("pi").unwrap(), 3.1415926535);
```
*/

#![warn(missing_docs)]

mod config;
pub use config::{Config, ConfigError};
